README.md
# uni

[![Build Status](https://semaphoreci.com/api/v1/sourcefoundry/uni/branches/master/badge.svg)](https://semaphoreci.com/sourcefoundry/uni) [![Build status](https://ci.appveyor.com/api/projects/status/1cbesa0qnbqgoj0p/branch/master?svg=true)](https://ci.appveyor.com/project/chrissimpkins/uni/branch/master) [![Go Report Card](https://goreportcard.com/badge/github.com/source-foundry/uni)](https://goreportcard.com/report/github.com/source-foundry/uni)

## About

`uni` is a command line application that performs glyph based searches for Unicode v10.0.0 defined code points and Unicode code point based searches for glyph shapes.

<img src="https://raw.githubusercontent.com/source-foundry/uni/img/img/uni1-crunch.png" alt="uni Argument Example" width="99%"/>

<img src="https://raw.githubusercontent.com/source-foundry/uni/img/img/uni3-crunch.png" alt="uni Line Filter Example" width="99%"/>

## Contents

- [Installation](#installation)
- [Usage](#usage)
- [Issues](#issues)
- [Contributing](#contributing)
- [License](#license)

## Installation

uni is developed in Go and compiled to the command line executable `uni` (`uni.exe` on Windows).  A variety of cross-compiled binaries are available for use on Linux, macOS, and Windows systems, or you can download the source and compile the application yourself.  Instructions for both approaches follow.

## Installation Approaches

### Approach 1: Install the pre-compiled binary executable file

Download the latest compiled release file for your operating system and architecture from [the Releases page](https://github.com/source-foundry/uni/releases/latest).

#### Linux / macOS

Unpack the tar.gz archive and move the `uni` executable file to a directory on your system PATH (e.g. `/usr/local/bin`).  This can be performed by executing the following command in the root of the unpacked archive:

```
$ mv uni /usr/local/bin/uni
```

There are no dependencies contained in the archive.  You can delete all downloaded archive files after the above step.

#### Windows

Unpack the zip archive and move the `uni.exe` executable file to a directory on your system PATH. See [details here](https://stackoverflow.com/questions/4822400/register-an-exe-so-you-can-run-it-from-any-command-line-in-windows) for more information about how to do this.

There are no dependencies contained in the archive.  You can delete all downloaded archive files after the above step.

### Approach 2: Compile from the source code and install

You must install the Go programming language (which includes the `go` tool) in order to compile the project from source.  Install Go v1.10 or above for Unicode v10.0.0 support.  Follow the [instructions on the Go download page](https://golang.org/dl/) for your platform. 

Once you have installed Go and configured your settings so that Go executables are installed on your system PATH, use the following command to (1) pull the master branch of the uni repository; (2) compile the `uni` executable from source for your platform/architecture configuration; (3) install the executable on your system:

```
$ go get github.com/source-foundry/uni
```

## Uninstall

The installation includes a single executable binary file.  If you installed with `go get` or added one of the pre-compiled binaries on your system `$PATH` on *.nix systems, you can uninstall with:

```
$ rm $(which uni)
```

## Usage

### Default behavior

By default, `uni` performs a search for Unicode code points with glyph arguments.  You can reverse this search behavior to identify glyphs with Unicode code point search strings by including the `-g` / `--glyph` option in your command.

### Options

```
  -g, --glyph          Search for glyph(s) with Unicode code point(s)
  -h, --help           Application help
      --usage          Application usage
  -v, --version        Application version
```

### Search for Unicode Code Points with Glyphs

#### Glyphs as arguments to `uni`

`uni` takes glyph arguments and displays the associated Unicode code points.  You can include the glyphs in a single string or separate them with spaces.  Use quotes around special shell characters.

```
$ uni [glyph 1]...[glyph n]
```

##### Example

```
$ uni Aa1Ø€βф▀र༩↵√ナ
U+0041 'A'
U+0061 'a'
U+0031 '1'
U+00D8 'Ø'
U+20AC '€'
U+03B2 'β'
U+0444 'ф'
U+2580 '▀'
U+0930 'र'
U+0F29 '༩'
U+21B5 '↵'
U+221A '√'
U+30CA 'ナ'
```

#### Glyphs piped through standard input stream

You can also pipe text data to `uni` through the standard input stream. `uni` will process every glyph that it receives in the stdin stream and print the associated Unicode code point to standard output.

```
$ [application command] | uni
```

#### Example

```
$ echo -n "Aa1Ø€βф▀र༩↵√ナ" | uni
U+0041 'A'
U+0061 'a'
U+0031 '1'
U+00D8 'Ø'
U+20AC '€'
U+03B2 'β'
U+0444 'ф'
U+2580 '▀'
U+0930 'र'
U+0F29 '༩'
U+21B5 '↵'
U+221A '√'
U+30CA 'ナ'
```

### Search for Glyphs with Unicode Code Points

Unicode code point searches are performed with the hexadecimal value.  Remove all prefix data from the string before you pass the request to uni. Include spaces between the Unicode code point arguments.

**Acceptable hexadecimal format**: `0041`

**Unacceptable hexadecimal formats**: `U+0041` `u+0041` `\u0041`

#### Unicode code points as arguments to `uni`

```
$ uni [Unicode code point 1]...[Unicode code point n]
```

##### Example

```
$ uni -g 0041 0031 03B2 2580 30CA
U+0041 'A'
U+0031 '1'
U+03B2 'β'
U+2580 '▀'
U+30CA 'ナ'
```

#### Unicode code points piped through standard input stream

```
$ [application command] | uni -g
```

##### Example

```
$ echo -n "0041 0031 03B2 2580 30CA" | uni
U+0041 'A'
U+0031 '1'
U+03B2 'β'
U+2580 '▀'
U+30CA 'ナ'
```

## Issues

Please [file an issue report](https://github.com/source-foundry/uni/issues/new) on the repository for any problems that arise with use.

## Contributing

Contributions to the project are encouraged and welcomed. Please base your work on the `dev` branch and submit pull requests with source changes.

## License

uni is licensed under the [MIT License](https://github.com/source-foundry/uni/blob/master/LICENSE).


CHANGELOG.md
## Changelog

### v2.0.0

- updated Unicode support to v10.0.  Adds 8,518 new Unicode code points.

### v1.0.0

- added support for Unicode code point to glyph search
- updated help and usage strings
- added Unicode Standard version report to version string

### v0.10.1

- fixed bug in goreleaser build script

### v0.10.0

- added line filter support - pipe test glyphs in through stdin stream
- refactored command line parsing source to an init function
- refactored validation function

### v0.9.0

- initial release

uni_test.go
package main

import (
	"bytes"
	"io"
	"log"
	"os"
	"regexp"
	"strings"
	"testing"
)

// Constant tests

func TestVersionString(t *testing.T) {
	r, _ := regexp.Compile(`\d{1,2}.\d{1,2}.\d{1,2}`)
	// match expected format of version string
	if r.MatchString(version) != true {
		t.Errorf("[FAIL] Unexpected version string format identified.")
	}
}

func TestUsageString(t *testing.T) {
	if strings.HasPrefix(usage, "Usage:") == false {
		t.Errorf("[FAIL] Unexpected usage string format.")
	}
}

func TestHelpString(t *testing.T) {
	if strings.HasPrefix(help, "=====") == false {
		t.Errorf("[FAIL] Unexpected help string format.")
	}
}

func TestMainFunctionSingleGlyph(t *testing.T) {
	os.Args = []string{"uni", "j"}
	old := os.Stdout // keep backup of stdout
	r, w, err := os.Pipe()
	if err != nil {
		log.Fatal(err)
	}
	os.Stdout = w

	outC := make(chan string)

	// copy the output in a separate goroutine
	go func() {
		var buf bytes.Buffer
		io.Copy(&buf, r)
		outC <- buf.String()
	}()

	main() // call main function with mock os.Args defined above

	// back to normal state
	w.Close()
	os.Stdout = old // restoring the real stdout
	out := <-outC

	if len(out) == 0 {
		t.Errorf("[FAIL] Test of main() function did not return standard output response")
	}

	if !(out == "U+006A 'j'\n") {
		t.Errorf("[FAIL] Expected execution of 'uni f' to return string U+006A 'j', but instead it returned %s", out)
	}

}

func TestMainFunctionMultiGlyph(t *testing.T) {
	os.Args = []string{"uni", "jj"}
	old := os.Stdout // keep backup of stdout
	r, w, err := os.Pipe()
	if err != nil {
		log.Fatal(err)
	}
	os.Stdout = w

	outC := make(chan string)

	// copy the output in a separate goroutine
	go func() {
		var buf bytes.Buffer
		io.Copy(&buf, r)
		outC <- buf.String()
	}()

	main() // call main function with mock os.Args defined above

	// back to normal state
	w.Close()
	os.Stdout = old // restoring the real stdout
	out := <-outC

	if len(out) == 0 {
		t.Errorf("[FAIL] Test of main() function did not return standard output response")
	}

	if !(out == "U+006A 'j'\nU+006A 'j'\n") {
		t.Errorf("[FAIL] Expected execution of 'uni f' to return string U+006A 'j' * 2, but instead it returned %s", out)
	}

}

func TestMainFunctionSingleUnicodeCodePoint(t *testing.T) {
	os.Args = []string{"uni", "-g", "006A"}
	old := os.Stdout // keep backup of stdout
	r, w, err := os.Pipe()
	if err != nil {
		log.Fatal(err)
	}
	os.Stdout = w

	outC := make(chan string)

	// copy the output in a separate goroutine
	go func() {
		var buf bytes.Buffer
		io.Copy(&buf, r)
		outC <- buf.String()
	}()

	main() // call main function with mock os.Args defined above

	// back to normal state
	w.Close()
	os.Stdout = old // restoring the real stdout
	out := <-outC

	if len(out) == 0 {
		t.Errorf("[FAIL] Test of main() function did not return standard output response")
	}

	if !(out == "U+006A 'j'\n") {
		t.Errorf("[FAIL] Expected execution of 'uni f' to return string U+006A 'j', but instead it returned %s", out)
	}

}

func TestMainFunctionMultiUnicodeCodePoint(t *testing.T) {
	os.Args = []string{"uni", "-g", "006A", "006A"}
	old := os.Stdout // keep backup of stdout
	r, w, err := os.Pipe()
	if err != nil {
		log.Fatal(err)
	}
	os.Stdout = w

	outC := make(chan string)

	// copy the output in a separate goroutine
	go func() {
		var buf bytes.Buffer
		io.Copy(&buf, r)
		outC <- buf.String()
	}()

	main() // call main function with mock os.Args defined above

	// back to normal state
	w.Close()
	os.Stdout = old // restoring the real stdout
	out := <-outC

	if len(out) == 0 {
		t.Errorf("[FAIL] Test of main() function did not return standard output response")
	}

	if !(out == "U+006A 'j'\nU+006A 'j'\n") {
		t.Errorf("[FAIL] Expected execution of 'uni f' to return string U+006A 'j' * 2, but instead it returned %s", out)
	}

}


search_test.go
package main

import (
	"strings"
	"testing"
)

// test single argument requests to unicodeSearch function for glyph --> Unicode code point search
func TestGlyphSearchCodePointsSingle(t *testing.T) {
	cases := []struct {
		unihex   string
		expected string
	}{
		{"006A", "U+006A 'j'"}, // ASCII
		{"20AC", "U+20AC '€'"}, // Currency
		{"03B2", "U+03B2 'β'"}, // Greek and Coptic
		{"0444", "U+0444 'ф'"}, // Cyrillic
		{"2580", "U+2580 '▀'"}, // Block elements
	}

	for _, c := range cases {
		response, err := glyphSearch(c.unihex)

		if err != nil {
			t.Errorf("[FAIL] Expected no error on glyph search and received %v", err)
		}

		if len(response) == 0 {
			t.Errorf("[FAIL] unicodeSearch did not return a value (len = 0)")
		} else if !strings.Contains(response, c.expected) {
			t.Errorf("[FAIL] Unicode code point '%s' yielded response %v, not expected response %s", c.unihex, response[0], c.expected)
		}
	}
}

// test single argument requests to unicodeSearch function for glyph --> Unicode code point search with appended newline
// character.  This is intended to address newline values that may come in through stdin piped text from another
// application.  Example is: `$ echo "006A" | uni -g` where newline is appended to the hexadecimal by the echo application
func TestGlyphSearchCodePointsSingleWithNewline(t *testing.T) {
	cases := []struct {
		unihex   string
		expected string
	}{
		{"006A\n", "U+006A 'j'"}, // ASCII
		{"20AC\n", "U+20AC '€'"}, // Currency
		{"03B2\n", "U+03B2 'β'"}, // Greek and Coptic
		{"0444\n", "U+0444 'ф'"}, // Cyrillic
		{"2580\n", "U+2580 '▀'"}, // Block elements
	}

	for _, c := range cases {
		response, err := glyphSearch(c.unihex)

		if err != nil {
			t.Errorf("[FAIL] Expected no error on glyph search and received %v", err)
		}

		if len(response) == 0 {
			t.Errorf("[FAIL] unicodeSearch did not return a value (len = 0)")
		} else if !strings.Contains(response, c.expected) {
			t.Errorf("[FAIL] Unicode code point '%s' yielded response %v, not expected response %s", c.unihex, response[0], c.expected)
		}
	}
}

func TestGlyphSearchInvalidHexadecimal(t *testing.T) {
	response, err := glyphSearch("zzzzz")
	if err == nil {
		t.Errorf("[FAIL] Expected error with invalid hexadecimal format and instead received response %s", response)
	}
}

func TestGlyphSearchInvalidOutOfRange(t *testing.T) {
	response, err := glyphSearch("0010")
	if err == nil {
		t.Errorf("[FAIL] Expected error with out of range hexadecimal format and instead received response %s", response)
	}
}

// test single argument requests to unicodeSearch function for glyph --> Unicode code point search
func TestUnicodeSearchCodePointsSingle(t *testing.T) {
	cases := []struct {
		glyph    string
		expected string
	}{
		{"j", "U+006A 'j'"}, // ASCII
		{"€", "U+20AC '€'"}, // Currency
		{"β", "U+03B2 'β'"}, // Greek and Coptic
		{"ф", "U+0444 'ф'"}, // Cyrillic
		{"▀", "U+2580 '▀'"}, // Block elements
	}

	for _, c := range cases {
		testglyph := []string{c.glyph}
		response := unicodeSearch(testglyph)

		if len(response) == 0 {
			t.Errorf("[FAIL] unicodeSearch did not return a value (len = 0)")
		} else if !strings.Contains(response[0], c.expected) {
			t.Errorf("[FAIL] Glyph '%s' yielded response %s, not expected response %s", c.glyph, response[0], c.expected)
		}
	}
}

// test multiple argument requests to the unicodeSearch function for glyph --> Unicode code point search
func TestUnicodeSearchCodePointsMultiple(t *testing.T) {
	cases := []struct {
		glyph    string
		expected string
	}{
		{"j", "U+006A 'j'"}, // ASCII
		{"€", "U+20AC '€'"}, // Currency
		{"β", "U+03B2 'β'"}, // Greek and Coptic
		{"ф", "U+0444 'ф'"}, // Cyrillic
		{"▀", "U+2580 '▀'"}, // Block elements
	}

	var testglyphs []string

	for _, c := range cases {
		testglyphs = append(testglyphs, c.glyph)
	}

	response := unicodeSearch(testglyphs)

	if len(response) == 0 {
		t.Errorf("[FAIL] unicodeSearch did not return a value (len = 0)")
	} else if !(len(response) == 5) {
		t.Errorf("[FAIL] Expected five Unicode points in response.  Received %d", len(response))
	}

	for i := range response {
		if !strings.Contains(response[i], cases[i].expected) {
			t.Errorf("[FAIL] Expected response %s but received response %s", cases[i].expected, response[i])
		}
	}
}


utilities_test.go
package main

import (
	"io/ioutil"
	"os"
	"strconv"
	"testing"
	"unicode/utf8"
)

func TestStdinValidatesTrueFunction(t *testing.T) {
	file, _ := ioutil.TempFile(os.TempDir(), "stdin")
	defer os.Remove(file.Name())

	file.WriteString("stdin test")

	result := stdinValidates(file)
	if result != true {
		t.Errorf("[FAIL] Attempt to validate mocked stdin failed.")
	}

}

func TestStdinValidatesFalseFunction(t *testing.T) {
	file, _ := ioutil.TempFile(os.TempDir(), "stdin")
	defer os.Remove(file.Name())

	file.WriteString("")

	result := stdinValidates(file)
	if result != false {
		t.Errorf("[FAIL] Attempt to validate empty mocked stdin failed.")
	}

}

func TestIsIntInRangeValid(t *testing.T) {
	needle, _ := strconv.ParseInt("00ff", 16, 32)
	ok := isIntInRange(int32(needle))
	if !ok {
		t.Errorf("[FAIL] Expected hexadecimal value U+00FF to be in range and received false")
	}
}

func TestIsIntInRangeTooLarge(t *testing.T) {
	needle := utf8.MaxRune + 1
	ok := isIntInRange(needle)
	if ok {
		t.Errorf("[FAIL] Expected value `utf8.MaxRun + 1` to be out of range and received true")
	}
}

func TestIsIntInRangeTooSmall(t *testing.T) {
	needle, _ := strconv.ParseInt("0019", 16, 32)
	ok := isIntInRange(int32(needle))
	if ok {
		t.Errorf("[FAIL] Expected hexadecimal value U+0019 to be out of range and received true")
	}
}


search.go
package main

import (
	"fmt"
	"strconv"
	"strings"
	"unicode/utf8"
)

// glyphSearch identifies glyphs for Unicode code point (hexadecimal string) search requests
func glyphSearch(arg string) (string, error) {
	// strip newline value if this was included (e.g. with stdin stream piped text from another application)
	if strings.Contains(arg, "\n") {
		arg = strings.Replace(arg, "\n", "", -1)
	}
	i, err := strconv.ParseInt(arg, 16, 32)
	if err != nil {
		return "", err
	}
	// test to confirm that Unicode code point falls in range U+0020 = min printable glyph to utf8.MaxRune value
	ok := isIntInRange(int32(i))
	if !ok {
		errmsg := fmt.Errorf("hexadecimal value '%s' was out of range", arg)
		return "", errmsg
	}
	r := rune(i)
	stdoutString := "U+" + arg + " '" + string(r) + "'"
	return stdoutString, nil
}

// unicodeSearch identifies the Unicode code point for glyph search requests
func unicodeSearch(argv []string) []string {
	var solist []string
	for i := 0; i < len(argv); i++ {
		if len(argv[i]) > 1 { // handle single argument that includes multiple glyphs
			charList := strings.Split(argv[i], "")
			for x := 0; x < len(charList); x++ {
				r, _ := utf8.DecodeRuneInString(charList[x])
				solist = append(solist, fmt.Sprintf("%#U", r))
			}
		} else { // handle multiple individual arguments
			r, _ := utf8.DecodeRuneInString(argv[i])
			solist = append(solist, fmt.Sprintf("%#U", r))
		}

	}
	return solist
}


utilities.go
package main

import (
	"os"
	"strconv"
	"unicode/utf8"
)

// test os.Stdin for presence of data, if present return true, else return false
func stdinValidates(stdin *os.File) bool {
	f, err := stdin.Stat()
	if err != nil { // unable to obtain file data with Stat() method call = fail
		return false
	}

	size := f.Size()
	if size == 0 { // there does not appear to be any data in the stdin stream = fail
		return false
	}

	return true
}

func isIntInRange(value int32) bool {
	minPrintableInt, _ := strconv.ParseInt("0020", 16, 32)

	if value > utf8.MaxRune {
		// value is higher than maximum acceptable Unicode code point
		return false
	}

	if value < int32(minPrintableInt) {
		// value is lower than minimum Unicode code point for printable glyphs
		return false
	}
	// return true if integer value falls within acceptable range
	return true
}


uni.go
// uni is a command line executable that performs Unicode code point to glyph and glyph to Unicode code point searches
package main

import (
	"bytes"
	"flag"
	"fmt"
	"io"
	"os"
	"strings"
	"unicode"
)

const (
	version = "2.0.0"

	usage = `Usage: uni (options) [arg 1]...[arg n]
Line Filter Usage: [application command] | uni (options)
`

	help = `=================================================
 uni
 Unicode v10.0 search tool
 Copyright 2018 Christopher Simpkins
 MIT License

 Source: https://github.com/source-foundry/uni
=================================================
 Usage:
  - With command line arguments:
      $ uni (options) [arg 1]...[arg n]
  - As line filter:
      $ [application command] | uni (options)
 Default:
   Search for Unicode code point(s) with glyph(s)
 Options:
  -g, --glyph          Search for glyph(s) with Unicode code point(s)
  -h, --help           Application help
      --usage          Application usage
  -v, --version        Application version
`
)

var versionShort, versionLong, helpShort, helpLong, usageLong, glyphShort, glyphLong *bool

//var listShort, listLong, listCommas, listNewLine *bool
var isGlyphSearch = false

func init() {
	// define available command line flags
	versionShort = flag.Bool("v", false, "Application version")
	versionLong = flag.Bool("version", false, "Application version")
	helpShort = flag.Bool("h", false, "Help")
	helpLong = flag.Bool("help", false, "Help")
	usageLong = flag.Bool("usage", false, "Usage")
	glyphShort = flag.Bool("g", false, "Glyph")
	glyphLong = flag.Bool("glyph", false, "Glyph")
	//listShort = flag.Bool("l", false, "List")
	//listLong = flag.Bool("list", false, "List")
	//listCommas = flag.Bool("comma", false, "Comma delimiters in lists")
	//listNewLine = flag.Bool("newline", false, "Newline delimiters in lists")
}

func main() {

	flag.Parse()

	// parse command line flags and handle them
	switch {
	case *versionShort, *versionLong:
		fmt.Printf("uni v%s\n", version)
		fmt.Printf("Unicode Standard v%s\n", unicode.Version)
		os.Exit(0)
	case *helpShort, *helpLong:
		fmt.Println(help)
		os.Exit(0)
	case *usageLong:
		fmt.Println(usage)
		os.Exit(0)
	}

	//if *listShort || *listLong {
	//	if len(os.Args) < 3 {
	//		os.Stderr.WriteString("[Error] Please include at least one Unicode code point range argument for the list.\n")
	//		os.Exit(1)
	//	}
	//	var finalUniList []string
	//	for _, x := range flag.Args() {
	//		if strings.Contains(x, "-") {
	//			tempUniList := strings.Split(x, "-")
	//			startInt, startErr := strconv.ParseInt(tempUniList[0], 16, 32)
	//			endInt, endErr := strconv.ParseInt(tempUniList[1], 16, 32)
	//			if startErr != nil {
	//				// TODO
	//			}
	//			if endErr != nil {
	//				// TODO
	//			}
	//			for i := startInt; i <= endInt; i++ {
	//				finalUniList = append(finalUniList, fmt.Sprintf("%U", i))
	//
	//			}
	//		}
	//	}
	//
	//	if *listCommas {
	//		fmt.Print(strings.Join(finalUniList, ","))
	//	} else if *listNewLine {
	//		fmt.Print(strings.Join(finalUniList, "\n"))
	//	} else {
	//		fmt.Print(strings.Join(finalUniList, " "))
	//	}
	//
	//	// TODO: need to exit to prevent code below from executing
	//	os.Exit(0)
	//}

	// create a flag for Unicode code point to glyph search command line requests
	if *glyphShort || *glyphLong {
		isGlyphSearch = true
	}

	// create a flag for stdin stream search based upon number of arguments and use of glyph search flag
	isStdinSearch := (!isGlyphSearch && len(os.Args) < 2) || (isGlyphSearch && len(os.Args) < 3)

	if isStdinSearch {
		if !stdinValidates(os.Stdin) {
			os.Stderr.WriteString("[Error] Please include at least one argument or pipe search requests to the executable through the stdin stream.\n")
			os.Stderr.WriteString(usage)
			os.Exit(1)
		}

		tmp := new(bytes.Buffer)
		if _, err := io.Copy(tmp, os.Stdin); err != nil {
			os.Stderr.WriteString("[Error] Failed to copy std input stream to memory. " + fmt.Sprintf("%v", err))
			os.Exit(1)
		}

		// stdin stream search for glyph from Unicode code point search request
		if isGlyphSearch {
			stdinList := strings.Split(tmp.String(), ` `)
			for _, arg := range stdinList {
				stdoutString, err := glyphSearch(arg)
				if err != nil {
					errmsg := fmt.Sprintf("%v", err)
					os.Stderr.WriteString("[Error] Unable to parse the Unicode code point request '" + arg + "' to a glyph. " + errmsg + "\n")
					os.Exit(1)
				}
				fmt.Println(stdoutString)
			}

		} else { // stdin stream search for Unicode code point from glyph search request
			stdinList := strings.Split(tmp.String(), "") // split the stdin string by glyph to a slice
			stdOutputList := unicodeSearch(stdinList)
			for _, stdoutString := range stdOutputList {
				fmt.Println(stdoutString)
			}
		}

	} else {
		// argument search for glyph from Unicode code point search request
		if isGlyphSearch {
			for _, arg := range os.Args[2:] {
				stdoutString, err := glyphSearch(arg)
				if err != nil {
					errmsg := fmt.Sprintf("%v", err)
					os.Stderr.WriteString("[Error] Unable to parse the Unicode code point request '" + arg + "' to a glyph. " + errmsg + "\n")
					os.Exit(1)
				}
				fmt.Println(stdoutString)
			}
		} else { // argument search for Unicode code point from glyph search request
			stdOutputList := unicodeSearch(os.Args[1:])
			for _, stdoutString := range stdOutputList {
				fmt.Println(stdoutString)
			}
		}
	}
}
